refactor: add more strict eslint rules

jonaslagoni opened the pull request:
**Description**
This PR adds a more strict eslint configuration to ensure we have consistency for developers to know exactly the syntax which is required. This is extended to test files as well.

**Related issue(s)**
Partly solves #190

magicmatatjahu commented:
@jonaslagoni I can accept the first two, but last `Force no curlies with for one lines curly` isn't good for me. I'm not a fan with that strict writing. I can accept code like `if (...) something()`, but

```ts
if (...)
  something()
```

isn't acceptable for me.

jonaslagoni commented:
> @jonaslagoni I can accept the first two, but last `Force no curlies with for one lines curly` isn't good for me. I'm not a fan with that strict writing. I can accept code like `if (...) something()`, but
> 
> ```ts
> if (...)
>   something()
> ```
> 
> isn't acceptable for me.

👍 Lets change it then...

jonaslagoni commented:
Soo, apperant the style we are looking for is not really [possible](https://stackoverflow.com/questions/65697539/prettier-forces-curlyless-single-line-if-statement-to-be-on-the-same-line-inste), I wonder if we want to force always brackets to enable consistancy 🤔?

magicmatatjahu commented:
> Soo, apperant the style we are looking for is not really possible, I wonder if we want to force always brackets to enable consistancy 🤔?

We can give free will to contributors in this topic :)

jonaslagoni commented:
> We can give free will to contributors in this topic :)

When another contributor comes along and uses 

```
if (...)
  something()
```

Would you then accept it? Cause if nothing is explicitly telling the developer that this syntax is not preferred it is IMO up to them to decide.

magicmatatjahu commented:
@jonaslagoni Yes, I forgot about that 😅 So maybe we should force the brackets, even for inline ifs

jonaslagoni commented:
> @jonaslagoni Yes, I forgot about that 😅 So maybe we should force the brackets for inline ifs

Yea I agree, lets do it.

magicmatatjahu reviewed (approved):
These inline returns look a bit weird 😅 I give accept to not block you anymore, but maybe we shouldn't bother with such rules? If someone will write inline ifs, then we should only aware about that, we'll just pay attention.

jonaslagoni commented:
> These inline returns look a bit weird 😅 I give accept to not block you anymore, but maybe we shouldn't bother with such rules? If someone will write inline ifs, then we should only aware about that, we'll just pay attention.

I would rather have us use one specific rule and force that, regardless of what it is, to be consistent. However, that is just my opinion, I can see it might be cumbersome when we are used to applying whatever style we see fit. 

In the long run, I think applying this rule is better so we don't mix it everywhere especially as each developer has their own opinion.

Don't mind this being blocked, as long as we agree on a decision in the end 🙂 

cc @czlowiek488

czlowiek488 commented:
It does not matter what is preffered by who but what the world standards are.  And you should not fight with those standards because they are standards for some reason.

You should not give a choice for developer, it should have strict autoformatting to make code consistent. Consistent code makes you read it faster because you know what you can expect in every part of codebase.

I would also recommend you to use prettier together with eslint. It may be a bit tricky to set it up so you may want to take a look at [this project](https://github.com/TheSoftwareHouse/express-boilerplate) and copy what you need. It may be also a source of good practices e.g. usage of dependency injection engine.

czlowiek488 commented on .eslintrc:
```diff
@@ -25,7 +26,8 @@
-    "curly": 0,
+    "no-loop-func": "error",
+    "curly": ["error"],
     "no-multi-spaces": 0,
```
You may want to add those rules.
    
    "require-await": "error",
    "no-only-tests/no-only-tests": "error",
    "no-console": "error",

jonaslagoni replied:
Added all but `no-only-tests/no-only-tests`, since IMO test should apply the same rules but a few.

jonaslagoni commented:
> It does not matter what is preffered by who but what the world standards are. And you should not fight with those standards because they are standards for some reason.
> 
> You should not give a choice for developer, it should have strict autoformatting to make code consistent. Consistent code makes you read it faster because you know what you can expect in every part of codebase.

I agree let's be strict in the syntax that can be used so the developer cannot be in doubt whether one or the other syntax should be used, and of course, so we are consistent.

I changed eslint to include tests now, and removed any unnecessary rules or changed them to "errors". The generic preset types @magicmatatjahu are going wild and I had to disable certain eslint rules for that file since the ripple effect was insane.

> I would also recommend you to use prettier together with eslint. It may be a bit tricky to set it up so you may want to take a look at [this project](https://github.com/TheSoftwareHouse/express-boilerplate) and copy what you need. It may be also a source of good practices e.g. usage of dependency injection engine.

@czlowiek488  long term that might be a great idea 👍 Feel free to create a feature request for this if you feel like it could be a good idea 🙂

czlowiek488 commented on src/generators/javascript/JavaScriptGenerator.ts:
```diff
@@ -27,17 +27,17 @@ export class JavaScriptGenerator extends AbstractGenerator<JavaScriptOptions> {
   }
 
-  async render(model: CommonModel, inputModel: CommonInputModel): Promise<string> {
+  render(model: CommonModel, inputModel: CommonInputModel): Promise<string> {
```
Shouldnt be here `: string` instead of `: Promise<string>`?

czlowiek488 replied:
I see that this is repeated in many more places.

jonaslagoni replied:
It is because we wanted to allow the async `presets` that we used `Promise` throughout 🙂

czlowiek488 commented on src/generators/java/renderers/EnumRenderer.ts:
```diff
@@ -34,7 +34,7 @@ ${this.indent(this.renderBlock(content, 2))}
   }
 
-  normalizeKey(value: any): string {
+  normalizeKey(value: unknown): string {
```
I do not think that unknown is a good here. I guess any was better. I think that unknown should be used only if you have API entrypoint and the you have unknown user input which is not used by your code at all.
I guess that you should do `any | number | boolean` and maybe bigint if it is possible.

jonaslagoni replied:
as soon as you have `any` what is the reason for adding more types? 🤔

czlowiek488 commented on test/utils/LoggingInterface.spec.ts:
```diff
@@ -94,10 +94,10 @@ describe('LoggingInterface', function() {
       expect(debugMock).toHaveBeenCalledTimes(1);
-      expect(debugMock).toHaveBeenCalledWith("test", {}, "test2");
+      expect(debugMock).toHaveBeenCalledWith('test', {}, 'test2');
       expect(infoMock).toHaveBeenCalledTimes(0);
```
This is overkill :D

jonaslagoni replied:
It is a simple `--fix` and I gotta admit I like that its constant 😄

jonaslagoni commented:
Because of this issue: https://github.com/asyncapi/.github/issues/55 I have to disable eslint rule for newlines.

magicmatatjahu reviewed (approved):
LGTM :)

jonaslagoni commented:
Thanks for the reviews @magicmatatjahu and @czlowiek488 👍
